File: app/src/main/java/com/rowland/delivery/features/auth/ui/AuthActivity.java
package com.rowland.delivery.features.auth.ui;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import com.google.android.material.textfield.TextInputLayout;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.rowland.delivery.features.auth.di.components.AuthComponent;
import com.rowland.delivery.features.auth.di.components.DaggerAuthComponent;
import com.rowland.delivery.features.dash.presentation.activities.DashActivity;
import com.rowland.delivery.merchant.R;
import com.rowland.delivery.merchant.application.di.modules.ContextModule;
import com.rowland.delivery.services.auth.Auth;
import com.rowland.delivery.services.auth.AuthException;
import com.rowland.delivery.services.auth.di.modules.AuthModule;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

import static com.rowland.delivery.services.auth.EmailAuth.CRED_EMAIL_KEY;
import static com.rowland.delivery.services.auth.EmailAuth.CRED_PASSWORD_KEY;

public class AuthActivity extends AppCompatActivity implements Auth.AuthLoginCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private Unbinder unbinder;

    @Inject
    @Named("google_login")
    Auth mGoogleAuth;

    @Inject
    @Named("email_login")
    Auth mEmailAuth;

    @BindView(R.id.splash_imageview)
    ImageView splash;

    @BindView(R.id.input_email)
    EditText input_email;

    @BindView(R.id.email_textinput_layout)
    TextInputLayout email_textinput_layout;

    @BindView(R.id.input_password)
    EditText input_password;

    @BindView(R.id.passord_textinput_layout)
    TextInputLayout passord_textinput_layout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_auth);

        unbinder = ButterKnife.bind(this);

        AuthComponent authComponent = DaggerAuthComponent.builder()
                .contextModule(new ContextModule(this))
                .authModule(new AuthModule())
                .build();

        authComponent.injectAuthActivity(this);

        Glide.with(this)
                .load(R.drawable.flash)
                .into(splash);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mGoogleAuth.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onLoginSuccess() {
        Toast.makeText(this, "Login Successful", Toast.LENGTH_SHORT).show();
        DashActivity.Companion.startActivity(this);
    }

    @Override
    public void onLoginFailure(AuthException e) {
        Toast.makeText(this, "Login Unsuccessful", Toast.LENGTH_SHORT).show();
    }

    @Override
    public Map<String, String> doEmailLogin() {
        Map<String, String> credentialsMap = new HashMap<>();
        credentialsMap.put(CRED_EMAIL_KEY, input_email.getText().toString());
        credentialsMap.put(CRED_PASSWORD_KEY, input_password.getText().toString());
        return credentialsMap;
    }

    @Override
    public Map<String, String> doEmailRegister() {
        Map<String, String> credentialsMap = new HashMap<>();
        credentialsMap.put(CRED_EMAIL_KEY, input_email.getText().toString());
        credentialsMap.put(CRED_PASSWORD_KEY, input_password.getText().toString());
        return credentialsMap;
    }

    public static void startActivity(Context context) {
        Intent intent = new Intent(context, AuthActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NO_HISTORY);
        context.startActivity(intent);

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
            Bundle bundle = ActivityOptions.makeSceneTransitionAnimation((Activity) context).toBundle();
            context.startActivity(intent, bundle);
        } else {
            context.startActivity(intent);
        }
    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {

    }

    @OnClick(R.id.btn_google_login)
    public void onGoogleLogin() {
        mGoogleAuth.login();
    }

    @OnClick(R.id.btn_login)
    public void onLogin() {
        mEmailAuth.login();
    }

    @OnClick(R.id.btn_register)
    public void onRegister() {
        mEmailAuth.register();
    }

    @OnClick(R.id.txt_reset_password)
    public void onPasswordReset() {
        mEmailAuth.login();
    }
}

File: ui/src/main/java/com/rowland/delivery/merchant/features/auth/ui/AuthActivity.kt
/*
 * Copyright (c) Otieno Rowland,  2021. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.rowland.delivery.merchant.features.auth.ui

import android.app.Activity
import android.app.ActivityOptions
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.bumptech.glide.Glide
import com.rowland.delivery.merchant.R
import com.rowland.delivery.merchant.databinding.ActivityAuthBinding
import com.rowland.delivery.merchant.features.auth.Auth
import com.rowland.delivery.merchant.features.auth.AuthException
import com.rowland.delivery.merchant.features.auth.EmailAuth
import com.rowland.delivery.merchant.features.dash.activities.DashActivity
import dagger.hilt.android.AndroidEntryPoint
import java.util.HashMap
import javax.inject.Inject
import javax.inject.Named

@AndroidEntryPoint
class AuthActivity : AppCompatActivity(), Auth.AuthLoginCallbacks {

    @Inject
    @Named("google_login")
    lateinit var mGoogleAuth: Auth

    @Inject
    @Named("email_login")
    lateinit var mEmailAuth: Auth

    private lateinit var binding: ActivityAuthBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAuthBinding.inflate(layoutInflater)
        setContentView(binding.root)

        Glide.with(this)
            .load(R.drawable.flash)
            .into(binding.splashImageview)

        binding.authContent.apply {
            btnGoogleLogin.setOnClickListener { mGoogleAuth.login() }
            btnLogin.setOnClickListener { mEmailAuth.login() }
            btnRegister.setOnClickListener { mEmailAuth.register() }
            txtResetPassword.setOnClickListener { mEmailAuth.login() }
        }
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        mGoogleAuth.onActivityResult(requestCode, resultCode, data!!)
    }

    override fun onLoginSuccess() {
        Toast.makeText(this, getString(R.string.login_successful), Toast.LENGTH_SHORT).show()
        DashActivity.startActivity(this)
    }

    override fun onLoginFailure(exception: AuthException) {
        Toast.makeText(this, getString(R.string.login_unsuccessful), Toast.LENGTH_SHORT).show()
        Log.e(AuthActivity::class.java.simpleName, "Login Unsuccessful", exception)
    }

    override fun doEmailLogin(): Map<String, String> {
        val credentialsMap = HashMap<String, String>()
        credentialsMap[EmailAuth.CRED_EMAIL_KEY] = binding.authContent.inputEmail.text.toString()
        credentialsMap[EmailAuth.CRED_PASSWORD_KEY] = binding.authContent.inputPassword.text.toString()
        return credentialsMap
    }

    override fun doEmailRegister(): Map<String, String> {
        val credentialsMap = HashMap<String, String>()
        credentialsMap[EmailAuth.CRED_EMAIL_KEY] = binding.authContent.inputEmail.text.toString()
        credentialsMap[EmailAuth.CRED_PASSWORD_KEY] = binding.authContent.inputPassword.text.toString()
        return credentialsMap
    }

    companion object {

        fun startActivity(context: Context) {
            val intent = Intent(context, AuthActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NO_HISTORY
            val bundle = ActivityOptions.makeSceneTransitionAnimation(context as Activity).toBundle()
            context.startActivity(intent, bundle)
            context.finish()
        }
    }
}